### PR TITLE
Enable JSON ASTs when 4th grammar section detected

### LINKED ISSUE #95: Enable JSON ASTs when 4th grammar section detected
Parent issue: https://github.com/ourPLCC/plcc/issues/52

Add an optional 4th section to grammar files. The 4th section will be populated by Python semantics. When a grammar files is written with four section, its structure is as follows:

lexical specification
%
syntactic specification
%
syntactic checks/helpers written in Java
%
semantics written in Python
Existing 3-section grammars should function normally.

When a 4th section is detected (even when empty), it will generate the Java code needed to generate JSON ASTs. Note that the 3rd section could also be empty, and it would behave the same.

### PR DESCRIPTION
**Pull request by StoneyJackson**
# Template for Final Squash Commit Message

```
refactor: enable Python when grammar has 4 sections

This is in anticipation of a new hybrid Java/Python mode
in which the parser runs in Java and the semantics rep-loop
runs in Python.

In this commit, when an optional 4th section is detected,
plcc.py generates ParseJsonAst.json. Thus allowing the
parser to to generate JSON ASTs when the `--json_ast`
flag is passed. It also generates experimental Python files.

---

Closes #95

Co-authored-by: Reed Everis <reednorrish@gmail.com>
Co-authored-by: Akshar Patel <aksharpatel1233@gmail.com>
Co-authored-by: Stoney Jackson <dr.stoney@gmail.com>
```

### CODE REVIEW

**Comment by StoneyJackson** [comment]
@reedeveris @reedeveris the build in CI has been fixed, and this branch has been synch'd.

**Comment by reedeveris** [comment]
Tentative to do:
	Add global flag for "python_semantics" in plccInit() lines 116-136
		(make the sempy routines separately toggled)
	Change 'destdir' flag in plccInit() to 'destdir_sem1'
	Add 'destdir_sem2' flag in plccInit() and set default to 'Python'
	Add statement to sem() to break for loop at '%', and remove 'done()' line 970
	Copy sem() into sempy() and refactor
	Copy semFinishUp() into sempyFinishUp() and refactor

**Comment by reedeveris** [comment]
As of now, plcc successfully detects the fourth section, and creates the Python directory, all automated based on whether the third section ends in a '%'.

BIG PROBLEM for @StoneyJackson 's attention:
@AksharP5 and I were discussing how to utilize the json functionality, but ran into some confusion. We noticed that currently, the json is generated in the command line when given a program. It's not making an AST, but a parse tree in json format. I don't believe this is the intended function? If it's not, we will need to make it so that it generates the **AST** in a json file (not the command line).

**Comment by reedeveris** [comment]
Note for next meeting:

  - Look at makeStub(), consider that in this method, you could potentially generate py files in given format.
  - Consider instead of 4 sections, 3 sections with the option to work in either Java or Python

**Comment by StoneyJackson** [comment]
@reedeveris @AksharP5 Please read https://github.com/ourPLCC/plcc/issues/52#issuecomment-1924629444

**Comment by StoneyJackson** [comment]
@reedeveris

When plcc.py processes a 4-section grammar, it will generate both Java and Python code. The Java code needs to include the ParseJsonAST.java file. Right now I think that's done by passing a --json_ast to plcc.py. We should change it so that it is done automatically if there is a 4th section.

I think all we should do in this PR is:

1. detect the forth section
2. copy ParseJsonAST.json to the Java/ directory
3. remove the --json_ast option (since all this did was (2), and now we are doing that when a 4th section is detected).

File: src/plcc.py [hunk 1]
```diff
@@ -924,9 +942,13 @@ def sem(nxt):
             argv.extend(fn)
             # print('== extend argv by {}'.format(fn))
             continue
+        if line == "%":
+            break
         if len(line) == 0 or line[0] == '#':
             # skip just comments or blank lines
             continue
+        if line == '%':
```
Comment (StoneyJackson) [inline src/plcc.py:950]: Why do we need this on both line 945 and line 950?

File: src/plcc.py [hunk 2]
```diff
@@ -1014,6 +1036,103 @@ def semFinishUp():
...
+
+
+def sempy(nxt):
+    global stubs, argv
+    # print('=== python semantic routines')
+    if not getFlag('python_semantics'):
+        sempyFinishUp()
+    for line in nxt:
+        line = line.strip()
+        if line[:7] == 'include':
```
Comment (StoneyJackson) [inline src/plcc.py:1048]: We're not going to allow the classic 'include' in the new 4th section. So let's remove this section.

File: tests/plcc/compiles-obj/OBJ/grammar [hunk 3]
```diff
@@ -157,3 +157,4 @@ include val
 include list
 include class
 include ref
+%
```
Comment (StoneyJackson) [inline tests/plcc/compiles-obj/OBJ/grammar:160]: We should leave this alone as it tests a traditional 3-section grammar. If you modify this, then we don't know if your changes break what was working before.

Instead, we need a separate test for a 4-section grammar. You could copy "compiles-obj" to something like "compiles-4-section-grammar", and then make this change.

**Review by StoneyJackson — COMMENTED** [review]
I haven't tried running this. But here is some feedback.

**Comment by StoneyJackson** [comment]
I see that there are updates. Excellent! I probably won't have a chance to look at them until Wednesday morning. So, if you are at the limits of what you can do in code without decisions from me, and you have more cycles for work, you can do some planning (next steps), design (Is there a design that would have made what you just did easer? Are the names the best they could be? How will the next step fit in with what you just did?), testing (write a test that passes when your new stuff works and fails when it doesn't), and/or documentation (how should README.md be updated to reflect the changes you just made).

**Comment by jashelio** [comment]
Perhaps I can be accused of being to much of an abstractionist, but here goes.

I feel that the following idea

> If there are three sections, then there is one section with all pre-execution checks and semantics in Java.
> If there are four sections, then there is a section in Java with pre-execution checks and another section in Python with semantics.

is short-sighted. I personally would prefer that we think in the following language-independent way.

> If there are three sections, then there is one section with all pre-execution checks and semantics.
> If there are four sections, then there is a section with pre-execution checks and another section with semantics.

And that we actually have options on the command line to say what languages are being used. Eventually we could change the section markers in the grammar file to say

% Java
 or
% Python

to get rid of those pesky command line options.

Just a thought,
Jim

**Comment by StoneyJackson** [comment]
@jashelio I don't understand your comment in the context of this PR. I think your comment is probably more appropriate for https://github.com/ourPLCC/plcc/issues/52 . Would you mind repeating yourself there, and then we can have the discussion there. Then I'll delete your last post and this post of mine.

FYI, I don't disagree with your premise, I just want to make sure it's in the context of the rest of the discussion of design and direction, and so we don't lose that line of discussion when this PR is done and merged.

**Comment by reedeveris** [comment]
Update:
  - Code injections should now be automatically normalized to a standard '\t' indent (this can be changed if we want)
Next: (notes for next meeting)
  - Java injection code vs Python injection code (we need both?)
  - Figure out testing
  - Ensure that the detection of 4th section enables JSON AST functionality

**Comment by reedeveris** [comment]
Create test cases:

  - no fourth section: Python directory should not be generated
  - empty fourth section: empty Python directory

**Pull request merged**